jumpscale/packages/vdc/billing.py

```diff
@@ -0,0 +1,143 @@
+    Args:
+        addon(obj): addon object 
+    Returns:
+        (float): addon price
+    """
+    prices = get_prices()
+    size = addon.size.name.lower()
+    return prices["nodes"][size]
+
+
+def calculate_plan_base_price():
+    """Get user plan price
+    Returns:
+        float(): the price of user plan
+    """
+    vdc_instance_name = list(j.sals.vdc.list_all())[0]
+    vdc_instance = j.sals.vdc.get(vdc_instance_name)
+    prices = get_prices()
+    return prices["plans"][vdc_instance.flavor.value]
+
+
+def calculate_addons_hourly_rate():
+    """Calcutlate the addons on hourly rate for all the addons by the user
+
+    Returns:
+        total_price (float): the total price for all addons
+    """
+
+    vdc_instance_name = list(j.sals.vdc.list_all())[0]
+    vdc_instance = j.sals.vdc.get(vdc_instance_name)
+    # addons = vdc_instance.addons
+    total_price = 0
+    # Calculate all the hourly late for all addons
+    addons = []
+    vdc_instance.load_info()
+    addons = get_addons(vdc_instance.flavor, vdc_instance.kubernetes)
+    for addon in addons:
+        addon_price = calculate_addon_price(addon)
+        j.logger.info(f"addon size {addon.size} with price {addon_price}")
+        total_price += addon_price / (24 * 30)
+    return total_price
+
+
+def calculate_hourly_rate():
+    """Calculate the total hourly rate of the user used plan and the addons.
+
+    Returns
+        hourly_amount(float): the total price for each hour,
+                              including the price of the user plan and addons
+    """
+    user_plan_price = calculate_plan_base_price()
+    hourly_amount = user_plan_price / (24 * 30)
+    j.logger.info(f"base plan price {user_plan_price} with hourly amount {hourly_amount}")
+    hourly_amount += calculate_addons_hourly_rate()
+    return hourly_amount
+
+
+def tranfer_prepaid_to_provision_wallet():
+    """Used to transfer the funds from prepaid wallet to provisioning wallet on an hourly basis
+    """
+    vdc_instance_name = list(j.sals.vdc.list_all())[0]
+    vdc_instance = j.sals.vdc.get(vdc_instance_name)
+    prepaid_wallet = vdc_instance.prepaid_wallet
+    provision_wallet = vdc_instance.provision_wallet
+    tft = prepaid_wallet.get_asset("TFT")
+    hourly_amount = calculate_hourly_rate()
+    j.logger.info(
+        f"starting the hourly transaction from prepaid wallet to provision wallet with total hourly amount {hourly_amount}"
+    )
+    hourly_amount = round(hourly_amount, 6)
+    prepaid_wallet.transfer(provision_wallet.address, hourly_amount, asset=f"{tft.code}:{tft.issuer}")
+
+
+def auto_extend_billing():
+    """Is used to get the pool in the VDC and extend them when the remaining time is less than
+    half of the BASE_CAPACITY
+    """
+    # Get the VDC and deployer instances
+
+    vdc_instance_name = list(j.sals.vdc.list_all())[0]
+    vdc_instance = j.sals.vdc.get(vdc_instance_name)
+    deployer = vdc_instance.get_deployer()
```

Comment:
`get_deployer` requires a password ?

Reply:
password is not required, if it is not passed, it will get the default identity